File: myLeetCode/Linked List/intersection-of-two-linked-lists.py
# 面试题 02.07. 链表相交
# https://leetcode.cn/problems/intersection-of-two-linked-lists-lcci/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if headA is None or headB is None:
            return None
        
        tailA = headA
        tailB = headB
        while tailA.next is not None:
            tailA = tailA.next
        while tailB.next is not None:
            tailB = tailB.next
        
        if tailA != tailB:
            return None
        
        p1 = headA
        p2 = headB
        while p1 != p2:
            p1 = headB if p1 is None else p1.next
            p2 = headA if p2 is None else p2.next
        
        return p1

def constructLinkedList(vals):
    head = None
    curr = None
    for val in vals:
        if head is None:
            head = ListNode(val)
            curr = head
        else:
            curr.next = ListNode(val)
            curr = curr.next
    return head

def main():
    A = [4, 1, 8, 4, 5]
    B = [5, 0, 1, 8, 4, 5]

    headA = constructLinkedList(A)
    headB = constructLinkedList(B)

    # 设置相交节点
    headB.next.next = headA.next.next

    sol = Solution()
    intersection = sol.getIntersectionNode(headA, headB)
    if intersection is not None:
        print(intersection.val)
    else:
        print("No intersection")

if __name__ == "__main__":
    main()

File: myLeetCode/Linked_List/intersection-of-two-linked-lists.cpp
// 面试题 02.07. 链表相交
// https://leetcode.cn/problems/intersection-of-two-linked-lists-lcci/

#include <iostream>
#include <vector>
struct ListNode {
    int val;
    ListNode *next;

    ListNode(int val) : val(val), next(nullptr) {}
    ListNode(int val, ListNode *next) : val(val), next(next) {}
};

class Solution {
  public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == nullptr || headB == nullptr) {
            return nullptr;
        }
        ListNode *tailA = headA;
        ListNode *tailB = headB;
        while (tailA->next != nullptr) {
            tailA = tailA->next;
        }
        while (tailB->next != nullptr) {
            tailB = tailB->next;
        }
        if (tailA != tailB) { // 两个链表不相交
            return nullptr;
        }
        ListNode *p1 = headA;
        ListNode *p2 = headB;
        while (p1 != p2) {
            p1 = (p1 == nullptr) ? headB : p1->next;
            p2 = (p2 == nullptr) ? headA : p2->next;
        }
        return p1; // 返回相交节点，如果没有相交节点则返回 nullptr。
    }
};

ListNode *constructLinkedList(std::vector<int> &vals) {
    ListNode *head = nullptr;
    ListNode *curr = nullptr;
    for (int val : vals) {
        if (head == nullptr) {
            head = new ListNode(val);
            curr = head;
        } else {
            curr->next = new ListNode(val);
            curr = curr->next;
        }
    }
    return head;
}

int main() {
    // std::vector<int> A = {0,9,1,2,4};
    // std::vector<int> B = {3,2,4};

    std::vector<int> A = {4, 1, 8, 4, 5};
    std::vector<int> B = {5, 0, 1, 8, 4, 5};

    ListNode *headA = constructLinkedList(A);
    ListNode *headB = constructLinkedList(B);

    // 设置相交节点
    headB->next->next = headA->next->next;

    Solution sol;
    ListNode *intersection = sol.getIntersectionNode(headA, headB);
    if (intersection != nullptr) {
        std::cout << intersection->val << std::endl;
    } else {
        std::cout << "No intersection" << std::endl;
    }
    return 0;
}